RadarSDK/RadarLogBuffer.m

```diff
@@ -35,19 +57,96 @@ + (instancetype)sharedInstance {
+        [mutableLogBuffer addObject:radarLog];
+
+        if (_persistentLogFeatureFlag || [[NSProcessInfo processInfo] environment][@"XCTestConfigurationFilePath"]) {
+            RadarLog *radarLog = [[RadarLog alloc] initWithLevel:level type:type message:message];
+            [inMemoryLogBuffer addObject:radarLog];
+            NSUInteger logLength = [inMemoryLogBuffer count];
+            if (logLength >= MAX_MEMORY_BUFFER_SIZE) {
+                [self persistLogs];
+            }
+        }   
+    }
+}
+
+- (void)persistLogs {
+    @synchronized (self) {
+        if (_persistentLogFeatureFlag || [[NSProcessInfo processInfo] environment][@"XCTestConfigurationFilePath"]) {
+            NSArray *flushableLogs = [inMemoryLogBuffer copy];
+            [self addLogsToBuffer:flushableLogs];
+            [inMemoryLogBuffer removeAllObjects]; 
+        }
+    }
+}
+
+- (NSArray<NSString *> *)getLogFilesInTimeOrder {
+    NSComparator compareTimeStamps = ^NSComparisonResult(NSString *str1, NSString *str2) {
+        return [@([str1 integerValue]) compare:@([str2 integerValue])];
+    };
+
+    return [self.fileHandler allFilesInDirectory:self.logFileDir withComparator:compareTimeStamps];
+}
+
+- (NSMutableArray<RadarLog *> *)readFromFileStorage {
+
+    NSArray<NSString *> *files = [self getLogFilesInTimeOrder];
+    NSMutableArray<RadarLog *> *logs = [NSMutableArray array];
+    if (!files) {
+        return logs;
+    }
+    for (NSString *file in files) {
+        NSString *filePath = [self.logFileDir stringByAppendingPathComponent:file];
+        NSData *fileData = [self.fileHandler readFileAtPath:filePath];
+        RadarLog *log = [NSKeyedUnarchiver unarchiveObjectWithData:fileData];
+        if (log && log.message) {
+            [logs addObject:log];
+        }
+    }
+
+    return logs;
+}
+
+ - (void)writeToFileStorage:(NSArray <RadarLog *> *)logs {
+    for (RadarLog *log in logs) {
+        NSData *logData = [NSKeyedArchiver archivedDataWithRootObject:log];
+        NSTimeInterval unixTimestamp = [log.createdAt timeIntervalSince1970];
+        //logs may be created in the same millisecond, so we append a counter to the end of the timestamp to "tiebreak"
+        NSString *unixTimestampString = [NSString stringWithFormat:@"%lld%04d", (long long)unixTimestamp, counter++];
+        NSString *filePath = [self.logFileDir stringByAppendingPathComponent:unixTimestampString];
+        [self.fileHandler writeData:logData toFileAtPath:filePath];
+    }
+ }
+
+- (void)append:(RadarLogLevel)level type:(RadarLogType)type message:(NSString *)message {
+    @synchronized (self) {
+        if (_persistentLogFeatureFlag || [[NSProcessInfo processInfo] environment][@"XCTestConfigurationFilePath"]) {
+            [self writeToFileStorage:@[[[RadarLog alloc] initWithLevel:level type:type message:message]]];
+        }
+        else{
+            [self write:level type:type message:message];
+        }
```

Comment:
I'm confused why we have both `write` and `append`. Let's just have one? With different logic based on feature flag?

Reply:
The write adds a log to the in memory buffer for batch writes and also can trigger a persist call which moves logs from memory to disk.

The append directly writes a log to disk and is needed for logging termination.

Reply:
Hmmm, I see.

I sorta think we want just another flag on `write` that's something like `forcePersistence` and then in the conditional where we choose to persist or not we do:
```
            if (logLength >= MAX_MEMORY_BUFFER_SIZE || forcePersist) {
                [self persistLogs];
            }
 ```
 
 `persistLogs` could be writing the in memory buffer from last to first so that we prioritize getting the termination event.
 
 I have a strong intuition we want one function here, not two.